chainerio/filesystems/posix.py

```diff
@@ -24,9 +24,18 @@ def open(self, file_path, mode='r',
                        buffering, encoding, errors,
                        newline, closefd, opener)
 
-    def list(self, path_or_prefix: str = None):
-        for file in os.scandir(path_or_prefix):
-            yield file.name
+    def list(self, path_or_prefix: str = None, recursive=False):
+        if recursive:
+            path_or_prefix.rstrip("/")
+            # use len instead of len + 1 as root in os.walk does not end
+            # with "/"
+            prefix_end_index = len(path_or_prefix)
+            for root, dirs, files in os.walk(path_or_prefix):
```

Comment:
Single directory listing can be very slow (e.g. overloaded NFS server) and thus I'd prefer recursively calling `os.scandir()` here.

Reply:
I think the `os.walk()` is implemented with `os.scandir()` recursively. 
https://github.com/python/cpython/blob/b9877cd2cc47b6f3512c171814c4f630286279b9/Lib/os.py#L350

Reply:
I checked the impl, but it's width-first and not depth-first, which would make us feel very slow when number of entries per single directory is huge, while our desired behaviour is depth-first.

Reply:
I am not sure I understand why depth-first is fast. My idea of implementing the depth-first would need a list of directories inside current directory to iterate into. While getting the list of directories needs to call `os.scandir` against current directory, which is the same as width-first.

Reply:
This  would be a depth-first listing code (and virtually done in hdfs filesystem list):
```python
def rec_list(path):
    for c in os.scandir(path):
        path1 = os.path.join(path, c.name)
        yield path1

        if c.isdir():
            yield from rec_list(path1)
```